services/inc/logging.h

```diff
@@ -0,0 +1,369 @@
+
+        if (LOG_ENABLED(TRACE)) {
+            // Do some heavy logging
+        }
+
+    LOG_MODULE_CATEGORY - defines logging category at module level. This macro is typically defined
+    via compiler arguments.
+
+    LOG_SOURCE_CATEGORY(name) - defines logging category for a source file (.c, .cpp). Source category
+    overrides module category.
+
+        #include "logging.h"
+        LOG_SOURCE_CATEGORY("foo.bar")
+
+    LOG_CATEGORY(name) - defines logging category specific to particular scope, such as function body
+    or class declaration. Scoped category overrides any global or currently visible category.
+
+        void baz() {
+            LOG_CATEGORY("foo.bar.baz");
+            ...
+        }
+
+    Following macros take category name as argument:
+
+    LOG_C(level, category, format, ...)
+    LOG_WRITE_C(level, category, data, size)
+    LOG_PRINT_C(level, category, str)
+    LOG_PRINTF_C(level, category, format, ...)
+    LOG_DUMP_C(level, category, data, size)
+    LOG_ENABLED_C(level, category)
+
+    Every logging macro has its debugging counterpart which is compiled only in debug builds:
+
+        LOG(INFO, "User name: %s", user);
+        LOG_DEBUG(INFO, "User password: %s", password);
+        LOG_DUMP(TRACE, public_key, sizeof(public_key));
+        LOG_DEBUG_DUMP(TRACE, private_key, sizeof(private_key));
+        ...
+
+    Following macros configure logging module at compile time:
+
+    LOG_INCLUDE_SOURCE_INFO - allows to include source info (file name, function name, line number)
+    into messages generated by the logging macros. By default, this macro is defined only for debug
+    builds (see config.h).
+
+    LOG_COMPILE_TIME_LEVEL - allows to strip any logging output that is below of certain logging level
+    at compile time. Default value is LOG_LEVEL_ALL meaning that no compile-time filtering is applied.
+
+    LOG_MAX_STRING_LENGTH - specifies maximum number of characters allowed for formatted strings.
+    This parameter affects log_message() and some other functions along with their wrapper macros.
+
+    LOG_DISABLE - disables logging entirely, turning all logging macros into no-op.
+*/
+
+#include <string.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include "panic.h"
+#include "config.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum LogLevel {
+    LOG_LEVEL_ALL = 1, // Log all messages
+    LOG_LEVEL_TRACE = 1,
+    LOG_LEVEL_INFO = 30,
+    LOG_LEVEL_WARN = 40,
+    LOG_LEVEL_ERROR = 50,
+    LOG_LEVEL_PANIC = 60,
+    LOG_LEVEL_NONE = 70, // Do not log any messages
+    // Compatibility levels
+    DEFAULT_LEVEL = 0,
+    ALL_LEVEL = LOG_LEVEL_ALL,
+    TRACE_LEVEL = LOG_LEVEL_TRACE,
+    LOG_LEVEL = LOG_LEVEL_TRACE,
+    DEBUG_LEVEL = LOG_LEVEL_TRACE,
+    INFO_LEVEL = LOG_LEVEL_INFO,
+    WARN_LEVEL = LOG_LEVEL_WARN,
+    ERROR_LEVEL = LOG_LEVEL_ERROR,
+    PANIC_LEVEL = LOG_LEVEL_PANIC,
+    NO_LOG_LEVEL = LOG_LEVEL_NONE
+} LogLevel;
+
+// Logger callbacks
```

Comment:
Could you add a description of the link between the logger callbacks and the log handler/log manager? It's not obvious why logging would need to callback to something else.

log_set_callbacks is the only link between spark_wiring_logging that lives in the user module and logging that lives in the HAL. From looking at both modules it seems the callbacks all define sinks for the logs, right?